risc0/zkvm/src/groth16.rs

```diff
@@ -0,0 +1,802 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Groth16 Verifier
+//!
+//! A library to verify Groth16 proofs computed over the BN_254 curve.
+//!
+//! # Example Usage
+//!
+//! ```rust
+//! use anyhow::Result;
+//! use risc0_zkvm::groth16::{
+//!     RawProof, RawPublic, RawVKey,
+//!     Groth16,
+//! };
+//!
+//! fn verification(circom_verification_key: &str, circom_proof: &str, circom_public: &str) -> Result<()> {
+//!     // parse the `verification_key`, `proof` and `public` as generated by Circom/SnarkJS
+//!     let raw_vkey: RawVKey = serde_json::from_str(circom_verification_key)?;
+//!     let raw_proof: RawProof = serde_json::from_str(circom_proof)?;
+//!     let raw_public = RawPublic {
+//!         values: serde_json::from_str(circom_public)?,
+//!     };
+//!
+//!     // build a groth16 instance from the raw material collected from Circom/SnarkJS
+//!     let groth16 = Groth16::from_raw(raw_vkey, raw_proof, raw_public)?;
+//!
+//!     // groth16 proof verification
+//!     groth16.verify()
+//! }
+//!
+//! ```
+
+use alloc::vec::Vec;
+#[cfg(not(target_os = "zkvm"))]
+use alloc::{format, string::String, vec};
+use anyhow::{anyhow, bail, Error, Result};
+use ark_bn254::{Bn254, G1Projective};
+#[cfg(not(target_os = "zkvm"))]
+use ark_bn254::{Fr, G1Affine, G2Affine};
+#[cfg(not(target_os = "zkvm"))]
+use ark_groth16::{prepare_verifying_key, VerifyingKey};
+use ark_groth16::{Groth16 as ark_Groth16, PreparedVerifyingKey, Proof};
+use ark_serialize::CanonicalDeserialize;
+#[cfg(not(target_os = "zkvm"))]
+use ark_serialize::CanonicalSerialize;
+#[cfg(not(target_os = "zkvm"))]
+use core::str::FromStr;
+#[cfg(not(target_os = "zkvm"))]
+use hex::FromHex;
+#[cfg(not(target_os = "zkvm"))]
+use num_bigint::BigInt;
+use serde::{Deserialize, Serialize};
+
+#[cfg(not(target_os = "zkvm"))]
+use crate::groth16::constants::*;
+#[cfg(not(target_os = "zkvm"))]
+use crate::sha::{Digest, Impl, Sha256};
+#[cfg(not(target_os = "zkvm"))]
+use crate::ALLOWED_IDS_ROOT;
+
+#[cfg(not(target_os = "zkvm"))]
+pub(crate) mod constants {
```

Comment:
How does this get updated? Is this entire file generated, or just these parameters?